lib/cog/adapters/config.ex

```diff
@@ -43,154 +49,215 @@ defmodule Cog.Adapters.Config do
-                            {:version, :hardcode, "v0.1"}],
-                 :numbers, [{:size, [:integer], :initial_size},
-                            {:max, [:required, :integer], :max_size}]]
-        
-        use Cog.Adapters.Config
+        use Cog.Adapters.Config,
+          key: Example.Config,
+          schema: [:strings, [{:name, [:required]},
+                              :description
+                              {:version, :hardcode, "v0.1"}],
+                   :numbers, [{:size, [:integer], :initial_size},
+                              {:max, [:required, :integer], :max_size}]]
       end
       
-      iex(1)> My.Adapter.ComplexConfig.fetch_config
+      iex(1)> My.Adapter.ComplexConfig.fetch_config!
       %{strings: %{name: "Example", decription: "My Example Configuration"
                    version: "v0.1"},
         numbers: %{size: 123, max: 200}}
-
   """
 
-  defmacro __using__(_) do
-    quote do
-      use GenServer
-      import Supervisor.Spec
-
-      @doc """
-      Reads configuration from Mix, applies validations and type coercion, and
-      returns a configuration hash. See Cog.Adapters.Config for more information.
-      """
-      def fetch_config() do
-        ensure_running
-        GenServer.call(__MODULE__, :get_config)
-      end
+  defmacro __using__(options) do
+    key = Keyword.get(options, :key, __CALLER__.module)
+    schema = Keyword.fetch!(options, :schema)
 
-      def fetch_config(group) do
-        Access.get(fetch_config, group)
-      end
+    quote location: :keep do
+      import Cog.Adapters.Config
+      require Logger
 
-      def start_link() do
-        GenServer.start_link(__MODULE__, [], name: __MODULE__)
+      def fetch_config do
+        Cog.Adapters.Config.fetch_config(unquote(key), unquote(schema))
       end
 
-      def init(_) do
-        {:ok, walk_config}
+      def fetch_config! do
+        Cog.Adapters.Config.fetch_config!(unquote(key), unquote(schema))
       end
+    end
+  end
 
-      def handle_call(:get_config, _from, state) do
-        {:reply, state, state}
-      end
+  # TODO: Alter the schema structure to deterministically identify namespaces
+  # from two tuple value definitions.
+  def fetch_config(key, schema) do
+    config = Application.get_env(:cog, key)
 
-      def walk_config(), do: walk_config(@schema)
-      def walk_config(schema) do
-        case Keyword.keyword?(schema) do
-          true ->
-            for key <- Keyword.keys(schema), into: %{},
-              do: {key, process_entries(Keyword.get(schema, key))}
-          false ->
-            process_entries(schema)
-        end
-      end
+    case Keyword.keyword?(schema) do
+      true ->
+        namespace_config(schema, config)
+      false ->
+        apply_schema(schema, config)
+    end
+  end
 
-      defp process_entries(entries) do
-        for entry <- entries, into: %{}, do: process_entry(entry)
-      end
+  def fetch_config!(key, schema) do
```

Comment:
I generally agree, unless there are lots of function heads that do totally different things. I typically remove whitespace for short one-liner funcs, but will add whitespace for larger implementations.